### DISCUSSION TITLE
🕹️ RetroType | Working on Sprite Animation MonkeyType Mod

### DISCUSSION THREAD

**Discussion by tcbutler320** [monkeytypegame/monkeytype#1506] [Ideas]
### *🕹️ RetroType*
> A MonkeyType modification to add pixel sprite animations to typing tests

https://user-images.githubusercontent.com/41709108/120866845-b76e7800-c55e-11eb-97b4-95d1a98fda5d.mov

#### *Background*   
I've been playing around with some game development projects and thought it would be fun to implement an animated sprite into MonkeyType. These videos represent what I've done so far.  

#### *Some Development Ideas*  
+ 👻 **Personalized Sprites** User's have option to choose their own avatar sprite. 
+ 🏆 **Trophy / Limited Edition Sprites** New sprites can be unlocked at certain levels, i.e high WPM scores, total races won, etc. 
 + 🎮 **New Game Modes** Updated game modes to take advantage of the sprite, some ideas include   
   +  **Race to the Flag** You have 15 seconds to get to the end of the flag (i.e, if you get there you typed ~100 WPM or something), gives you some extra motivation to try to get to the goal WPM
   +  **Fight Mode** This would be intense, but it might be possible to set up some kind of sprite fight sequence where correct words in a row = special combo's, total WPM = damage dealt, etc

#### *Current Sprite Options* 

https://user-images.githubusercontent.com/41709108/120755929-99f8ca00-c4dc-11eb-9b28-f418f4de14f7.mp4

#### *💾 Technicalities*  
I used event listeners for the sprite animations. You can enter `free move` mode by just using the nav arrows to walk around. When you start a typing test, pressing alphanumeric keys will move the sprite to the right by 1 space, backspace will move the sprite left by 1 space. 

#### *😰 Concerns*  
Animating a local sprite with eventListeners is pretty trivial, but I haven't yet looked into how this could be done in multiplayer mode. It might be possible to work this into the current `socket.io` socket scheme, or there could be some javascript game library. To look as good as it does locally, the sockets would have to update per each user per each keystroke, which would be a crazy increase in socket messages. Either way, if this were to become a serious development idea then it might be advantageous to look into how resource intensive this would be.

**Comment by Miodec** [comment]
I dont know, i feel like this is really out there and doesnt fully fit the look of the website. Plus im not sure i understand the functionality. The sprite just moves right as you type?

**Reply by tcbutler320** [reply]
Yeah there are a couple sites that do something similar. I think they all look a bit clunky but I agree its a little different then the current site's look. I'm totally impartial to building this out, just currently using this to explore sprite animations in general 

For functionality,the idea is like the sprite gets to the end or opens the chest if you hit your target.I.e the sprite only gets to the chest on 10 word mode if you get 100 WPM or something. It could change to be you must get a new PR to hit the chest. Could change this to different monkey avatar sprites and bananas too.


https://user-images.githubusercontent.com/41709108/120864227-e20a0200-c559-11eb-87ad-d6f47baa277d.mov
